### PR TITLE
Improve SSLOnlineEvaluator

### PR DESCRIPTION
**Pull request by zhengyu-yang**
## What does this PR do?

Make `SSLOnlineEvaluator` supports DDP and checkpoint.

1. Current `SSLOnlineEvaluator` does not support DDP. Weights of classification layers are not synced across different processes.
2. Current `SSLOnlineEvaluator` will break `resume_from_checkpoint` functionality of `Trainer` because `SSLOnlineEvaluator` add an additional `non_linear_evaluator` to the main `pl_module` during `on_pretrain_routine_start`. This will cause the statedict to mismatch when `resume_from_checkpoint`.
3. Both `num_classes` and `dataset` arguments can be inferred during runtime. We do not need to pass them in during initialization.
4. When train the online evaluator, the main `pl_module` should use `eval` mode.

This pull request deals with the above mentioned issues. 

## Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? \[not needed for typos/docs\]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-bolts/blob/master/CHANGELOG.md)?

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)

### CODE REVIEW

**Comment by zhengyu-yang** [comment]
Need to add `convert_sync_batchnorm` in case a non-linear online evaluator is used.

File: pl_bolts/callbacks/ssl_online.py [hunk 1]
```diff
@@ -1,74 +1,96 @@
...
-        self.optimizer: Optimizer
 
-        self.z_dim = z_dim
-        self.num_classes = num_classes
-        self.dataset = dataset
+        self.optimizer: Optional[Optimizer] = None
+        self.online_evaluator: Optional[SSLEvaluator] = None
+        self.z_dim: Optional[int] = None
+        self.num_classes: Optional[int] = None
+        self.dataset: Optional[str] = None
```
Comment (Borda) [inline pl_bolts/callbacks/ssl_online.py:56]: how these are set?

File: pl_bolts/callbacks/ssl_online.py [hunk 2]
```diff
@@ -1,74 +1,96 @@
...
         self.hidden_dim = hidden_dim
         self.drop_p = drop_p
-        self.optimizer: Optimizer
 
-        self.z_dim = z_dim
-        self.num_classes = num_classes
-        self.dataset = dataset
+        self.optimizer: Optional[Optimizer] = None
+        self.online_evaluator: Optional[SSLEvaluator] = None
+        self.z_dim: Optional[int] = None
```
Comment (Borda) [inline pl_bolts/callbacks/ssl_online.py:54]: seems overwrite line 48

File: pl_bolts/callbacks/ssl_online.py (hunk 1 above)
Comment (zhengyu-yang) [reply pl_bolts/callbacks/ssl_online.py:56]: They are set during `setup()`

File: pl_bolts/callbacks/ssl_online.py (hunk 2 above)
Comment (zhengyu-yang) [reply pl_bolts/callbacks/ssl_online.py:54]: I forgot to remove line 54.

File: pl_bolts/callbacks/ssl_online.py (hunk 1 above)
Comment (zhengyu-yang) [reply pl_bolts/callbacks/ssl_online.py:55]: See [here](https://github.com/zhengyu-yang/lightning-bolts/blob/edbfca7ee9a283df2d9aee3e8705cb1c53bffe3f/pl_bolts/callbacks/ssl_online.py#L59-L61)

Comment (Borda) [reply pl_bolts/callbacks/ssl_online.py:55]: well, can we still keep the init arguments and if needed set default as None?
or pls what is the reason for removing these arguments from init?

Comment (zhengyu-yang) [reply pl_bolts/callbacks/ssl_online.py:55]: This is more of a design choice. As this is adapted from my research code, and in my experiments, I try to remove any redundancy in configuration. 
I can add them back for backward compatibility.

Comment (Borda) [reply pl_bolts/callbacks/ssl_online.py:55]: > I can add them back for backward compatibility. Do you think it is necessary?

yes, pls add it back as it does not hurt you and we do not break any user code...

Comment (SeanNaren) [reply pl_bolts/callbacks/ssl_online.py:55]: Thanks @zhengyu-yang as @Borda suggested, it would be nice to keep the signature the same for BW compat :)

**Review by SeanNaren — COMMENTED** [review]
Great work @zhengyu-yang! Everything else looks good to me, however am curious if the online evaluator works well considering the entire module is wrapped in DDP, and then the online evaluator is also wrapped.

**Comment by Borda** [comment]
@zhengyu-yang @ananyahjha93 any progress here? :rabbit:

**Comment by zhengyu-yang** [comment]
I have made `num_classes ` and `dataset ` optional

**Comment by Borda** [comment]
@zhengyu-yang @SeanNaren any update here? :rabbit:

**Comment by zhengyu-yang** [comment]
> @zhengyu-yang @SeanNaren any update here? 🐰
Is there any todo on my side?

**Comment by Borda** [comment]
> Is there any todo on my side?

seems that the GPU tests are failing...

**Comment by ant0nsc** [comment]
@zhengyu-yang thanks very much for this super helpful change, we need that right now, too!
Just one question around how you designed that: Why are you not loading the state dict of the model/optimizer directly in `on_load_checkpoint`? Rather, you are saving it in a private member, and only use it in `on_pretrain_routine_start`. Why is that?

**Pull request merged**